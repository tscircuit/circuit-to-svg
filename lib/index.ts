export * from "./pcb/convert-circuit-json-to-pcb-svg"
export * from "./assembly/convert-circuit-json-to-assembly-svg"
export * from "./pinout/convert-circuit-json-to-pinout-svg"
export * from "./sch/convert-circuit-json-to-schematic-svg"
export * from "./pcb/convert-circuit-json-to-solder-paste-mask"
export type { ColorMap } from "./utils/colors"
export type { PcbColorMap, PcbColorOverrides } from "./pcb/colors"
export { getSoftwareUsedString } from "./utils/get-software-used-string"
export { CIRCUIT_TO_SVG_VERSION } from "./package-version"
export { createSvgObjectsForSchComponentPortHovers } from "./sch/svg-object-fns/create-svg-objects-for-sch-port-hover"
