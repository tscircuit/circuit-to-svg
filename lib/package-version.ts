import pkg from "../package.json" with { type: "json" }

export const CIRCUIT_TO_SVG_VERSION = (pkg as { version: string }).version
