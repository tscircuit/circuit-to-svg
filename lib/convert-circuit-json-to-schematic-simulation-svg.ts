import type { AnyCircuitElement } from "circuit-json"
import { stringify, parseSync } from "svgson"
import { convertCircuitJsonToSchematicSvg } from "./sch/convert-circuit-json-to-schematic-svg"
import { convertCircuitJsonToSimulationGraphSvg } from "./sim/convert-circuit-json-to-simulation-graph-svg"
import {
  type CircuitJsonWithSimulation,
  isSimulationExperiment,
  isSimulationTransientVoltageGraph,
} from "./sim/types"
import { CIRCUIT_TO_SVG_VERSION } from "./package-version"
import { getSoftwareUsedString } from "./utils/get-software-used-string"
import type { SvgObject } from "./svg-object"

interface ConvertSchematicSimulationParams {
  circuitJson: CircuitJsonWithSimulation[]
  simulation_experiment_id: string
  simulation_transient_voltage_graph_ids?: string[]
  width?: number
  height?: number
  schematicHeightRatio?: number
  schematicOptions?: Omit<
    Parameters<typeof convertCircuitJsonToSchematicSvg>[1],
    "width" | "height" | "includeVersion"
  >
  includeVersion?: boolean
}

const DEFAULT_WIDTH = 1200
const DEFAULT_HEIGHT = 1200
const DEFAULT_SCHEMATIC_RATIO = 0.55

export function convertCircuitJsonToSchematicSimulationSvg({
  circuitJson,
  simulation_experiment_id,
  simulation_transient_voltage_graph_ids,
  width = DEFAULT_WIDTH,
  height = DEFAULT_HEIGHT,
  schematicHeightRatio = DEFAULT_SCHEMATIC_RATIO,
  schematicOptions,
  includeVersion,
}: ConvertSchematicSimulationParams): string {
  const schematicElements = circuitJson.filter(
    (element): element is AnyCircuitElement =>
      !isSimulationExperiment(element) &&
      !isSimulationTransientVoltageGraph(element),
  )

  const clampedRatio = clamp01(schematicHeightRatio)
  const rawSchematicHeight = Math.max(1, height * clampedRatio)
  const rawSimulationHeight = Math.max(1, height - rawSchematicHeight)
  const totalRawHeight = rawSchematicHeight + rawSimulationHeight
  const scale = totalRawHeight === 0 ? 1 : height / totalRawHeight
  const schematicHeight = rawSchematicHeight * scale
  const simulationHeight = rawSimulationHeight * scale

  const schematicSvg = convertCircuitJsonToSchematicSvg(schematicElements, {
    ...schematicOptions,
    width,
    height: schematicHeight,
    includeVersion: false,
  })

  const simulationSvg = convertCircuitJsonToSimulationGraphSvg({
    circuitJson,
    simulation_experiment_id,
    simulation_transient_voltage_graph_ids,
    width,
    height: simulationHeight,
    includeVersion: false,
  })

  const schematicNode = ensureElementNode(parseSync(schematicSvg))
  const simulationNode = ensureElementNode(parseSync(simulationSvg))

  const combinedChildren: SvgObject[] = []
  combinedChildren.push(
    translateNestedSvg(schematicNode, 0, 0, width, schematicHeight),
  )
  combinedChildren.push(
    translateNestedSvg(
      simulationNode,
      0,
      schematicHeight,
      width,
      simulationHeight,
    ),
  )

  const softwareUsedString = getSoftwareUsedString(schematicElements)

  const svgObject: SvgObject = {
    name: "svg",
    type: "element",
    value: "",
    attributes: {
      xmlns: "http://www.w3.org/2000/svg",
      width: formatNumber(width),
      height: formatNumber(height),
      viewBox: `0 0 ${formatNumber(width)} ${formatNumber(height)}`,
      "data-simulation-experiment-id": simulation_experiment_id,
      ...(softwareUsedString && {
        "data-software-used-string": softwareUsedString,
      }),
      ...(includeVersion && {
        "data-circuit-to-svg-version": CIRCUIT_TO_SVG_VERSION,
      }),
    },
    children: combinedChildren,
  }

  return stringify(svgObject)
}

function translateNestedSvg(
  node: SvgObject,
  x: number,
  y: number,
  width: number,
  height: number,
): SvgObject {
  const clone = cloneSvgObject(node)
  clone.attributes = {
    ...clone.attributes,
    x: formatNumber(x),
    y: formatNumber(y),
    width: formatNumber(width),
    height: formatNumber(height),
  }

  delete clone.attributes.xmlns
  return clone
}

function ensureElementNode(node: SvgObject): SvgObject {
  if (node.type !== "element") {
    throw new Error("Expected SVG root element to be of type 'element'")
  }
  return node
}

function cloneSvgObject(node: SvgObject): SvgObject {
  return {
    ...node,
    attributes: { ...(node.attributes ?? {}) },
    children: node.children?.map(cloneSvgObject) ?? [],
  }
}

function clamp01(value: number): number {
  if (!Number.isFinite(value)) return DEFAULT_SCHEMATIC_RATIO
  if (value <= 0) return 0
  if (value >= 1) return 1
  return value
}

function formatNumber(value: number): string {
  if (!Number.isFinite(value)) return "0"
  const rounded = Number.parseFloat(value.toFixed(6))
  if (Number.isInteger(rounded)) return rounded.toString()
  return rounded.toString()
}
