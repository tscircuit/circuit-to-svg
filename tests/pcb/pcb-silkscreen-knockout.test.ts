// @ts-nocheck
import { test, expect } from "bun:test";
import { writeFileSync } from "fs";
import { convertCircuitJsonToPcbSvg } from "lib";

test("pcb silkscreen text (no knockout)", () => {
  const svg = convertCircuitJsonToPcbSvg([
    {
      type: "pcb_silkscreen_text",
      pcb_component_id: "pcb_silkscreen_text_0",
      layer: "top",
      anchor_position: { x: 10, y: 10 },
      anchor_alignment: "center",
      text: "HV-IN",
      font_size: 2.4,
      ccw_rotation: 0,
      knockout: false,
      knockout_padding: 0.25,
    },
  ] as any);

  // snapshot z bun/test
  expect(svg).toMatchSnapshot();
});

test("pcb silkscreen text (knockout + padding)", () => {
  const svg = convertCircuitJsonToPcbSvg([
    {
      type: "pcb_silkscreen_text",
      pcb_component_id: "pcb_silkscreen_text_1",
      layer: "top",
      anchor_position: { x: 20, y: 12 },
      anchor_alignment: "center",
      text: "HELLO",
      font_size: 2.4,
      ccw_rotation: 0,
      knockout: true,
      knockout_padding: 0.5,
    },
  ] as any);

  // ustvarimo datoteko za PR screenshot
  writeFileSync("knockout-demo.svg", svg);

  // vsaj preveri, da je maska v SVG-ju
  expect(svg).toContain("<mask");
  expect(svg).toMatchSnapshot();
});

test("pcb silkscreen text (knockout rotated 15Â°)", () => {
  const svg = convertCircuitJsonToPcbSvg([
    {
      type: "pcb_silkscreen_text",
      pcb_component_id: "pcb_silkscreen_text_2",
      layer: "top",
      anchor_position: { x: 30, y: 15 },
      anchor_alignment: "center",
      text: "VIN",
      font_size: 2.0,
      ccw_rotation: 15,
      knockout: true,
      knockout_padding: 0.3,
    },
  ] as any);

  expect(svg).toContain("<mask");
  expect(svg).toMatchSnapshot();
});
