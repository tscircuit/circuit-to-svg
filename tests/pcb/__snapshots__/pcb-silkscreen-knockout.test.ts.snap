// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`pcb silkscreen text (no knockout) 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="800" height="600"><style></style><rect class="boundary" x="0" y="0" fill="#000" width="800" height="600"/><rect class="pcb-boundary" fill="none" stroke="#fff" stroke-width="0.3" x="400" y="300" width="0" height="0"/><text x="0" y="0" dx="0" dy="0" fill="#f2eda1" font-family="Arial, sans-serif" font-size="720" text-anchor="middle" dominant-baseline="central" transform="matrix(1,0,0,1,400,300)" class="pcb-silkscreen-text pcb-silkscreen-top" data-pcb-silkscreen-text-id="pcb_silkscreen_text_0" stroke="none">HV-IN</text></svg>"`;

exports[`pcb silkscreen text (knockout + padding) 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="800" height="600"><style></style><rect class="boundary" x="0" y="0" fill="#000" width="800" height="600"/><rect class="pcb-boundary" fill="none" stroke="#fff" stroke-width="0.3" x="400" y="300" width="0" height="0"/><g transform="rotate(0, 400, 300)"><defs><mask id="pcb-silk-ko-1"><rect x="-1230" y="-582" width="2460" height="1164" fill="white" transform="matrix(1,0,0,1,400,300)"/><text x="0" y="0" dx="0" dy="0" fill="black" font-family="Arial, sans-serif" font-size="720" text-anchor="middle" dominant-baseline="central" transform="matrix(1,0,0,1,400,300)">HELLO</text></mask></defs><rect x="-1230" y="-582" width="2460" height="1164" fill="#f2eda1" transform="matrix(1,0,0,1,400,300)" mask="url(#pcb-silk-ko-1)" class="pcb-silkscreen-text pcb-silkscreen-top" data-pcb-silkscreen-text-id="pcb_silkscreen_text_1"/></g></svg>"`;

exports[`pcb silkscreen text (knockout rotated 15Â°) 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="800" height="600"><style></style><rect class="boundary" x="0" y="0" fill="#000" width="800" height="600"/><rect class="pcb-boundary" fill="none" stroke="#fff" stroke-width="0.3" x="400" y="300" width="0" height="0"/><g transform="rotate(15, 400, 300)"><defs><mask id="pcb-silk-ko-2"><rect x="-630" y="-450" width="1260" height="900" fill="white" transform="matrix(1,0,0,1,400,300)"/><text x="0" y="0" dx="0" dy="0" fill="black" font-family="Arial, sans-serif" font-size="600" text-anchor="middle" dominant-baseline="central" transform="matrix(0.9659258262890683,0.25881904510252074,-0.25881904510252074,0.9659258262890683,400,300)">VIN</text></mask></defs><rect x="-630" y="-450" width="1260" height="900" fill="#f2eda1" transform="matrix(1,0,0,1,400,300)" mask="url(#pcb-silk-ko-2)" class="pcb-silkscreen-text pcb-silkscreen-top" data-pcb-silkscreen-text-id="pcb_silkscreen_text_2"/></g></svg>"`;
